File: demo/java/A2ADemo.java
/**
 * mq9 A2A Demo
 * ============
 * Two-way A2A communication over mq9: Agent B sends a task to Agent A
 * (a translator) and receives the result back via its own mailbox.
 *
 * Run two terminals:
 *
 *   Terminal 1 — start Agent A (translator):
 *     mvn compile exec:java -Dexec.mainClass=A2ADemo -Dexec.args=agent
 *
 *   Terminal 2 — run Agent B (sender):
 *     mvn compile exec:java -Dexec.mainClass=A2ADemo -Dexec.args=client
 *
 * Or run both in the same JVM:
 *     mvn compile exec:java -Dexec.mainClass=A2ADemo
 */

import io.a2a.spec.Message;
import io.a2a.spec.MessageSendParams;
import io.a2a.spec.SendMessageRequest;
import io.a2a.spec.TextPart;
import io.mq9.ConsumeOptions;
import io.mq9.a2a.A2AContext;
import io.mq9.a2a.EventQueue;
import io.mq9.a2a.Mq9A2AAgent;

import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public class A2ADemo {

    private static final String SERVER =
            System.getenv().getOrDefault("MQ9_SERVER", "nats://demo.robustmq.com:4222");

    // ── Agent A — translator ───────────────────────────────────────────────────

    static void runAgent() throws Exception {
        Mq9A2AAgent agentA = Mq9A2AAgent.builder().server(SERVER).build();

        // A2A protocol event sequence: WORKING → Artifact → COMPLETED
        agentA.onMessage(
            (A2AContext ctx, EventQueue queue) ->
                // A2A protocol: WORKING first — tells the sender processing has started
                queue.working(ctx)
                    .thenCompose(v -> {
                        // A2A protocol: message body is one or more Parts; get the first text
                        String text = ctx.firstTextPart().orElse("");
                        String translated = "[translated] " + text;
                        System.out.println("[agent-a] '" + text + "' → '" + translated + "'");
                        // A2A protocol: push result as Artifact — call multiple times for streaming
                        return queue.artifact(ctx, "translation", translated);
                    })
                    // A2A protocol: COMPLETED last — signals the task is done
                    .thenCompose(v -> queue.completed(ctx)),
            ConsumeOptions.builder()
                .groupName("demo.java.translator.workers")
                .deliver("earliest")
                .numMsgs(10)
                .maxWaitMs(500)
                .build()
        );

        agentA.connect().join();
        String mailbox = agentA.createMailbox("demo.java.translator", 0).join();
        System.out.println("[agent-a] mailbox=" + mailbox);
        System.out.println("[agent-a] running — waiting for tasks (Ctrl+C to stop)");

        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            agentA.unregister().join();
            agentA.close();
            System.out.println("[agent-a] shut down");
        }));
        Thread.currentThread().join();
    }

    // ── Agent B — sender + receiver ────────────────────────────────────────────

    static void runClient() throws Exception {
        Mq9A2AAgent agentB = Mq9A2AAgent.builder().server(SERVER).build();

        CountDownLatch done = new CountDownLatch(1);

        // All messages arrive here — both reply events and new incoming tasks.
        // Use context.taskId to tell them apart: if it matches a task you sent,
        // it's a reply; otherwise it's a new task from another agent.
        agentB.onMessage(
            (A2AContext ctx, EventQueue queue) -> {
                System.out.println("[agent-b] reply event  task_id=" + ctx.taskId);
                ctx.firstTextPart().ifPresent(t -> System.out.println("[agent-b] text: " + t));
                done.countDown();
                return CompletableFuture.completedFuture(null);
            },
            ConsumeOptions.builder()
                .groupName("demo.java.sender.workers")
                .deliver("earliest")
                .numMsgs(10)
                .maxWaitMs(500)
                .build()
        );

        agentB.connect().join();
        String bMailbox = agentB.createMailbox("demo.java.sender", 300).join();
        System.out.println("[agent-b] mailbox=" + bMailbox);

        // Discover Agent A
        System.out.println("[agent-b] discovering translator agents…");
        List<Map<String, Object>> agents = agentB.discover("translation", false, 5).join();
        if (agents.isEmpty()) {
            System.out.println("[agent-b] no agents found — is agent-a running?");
            agentB.close();
            return;
        }
        Map<String, Object> target = agents.get(0);
        System.out.println("[agent-b] found: name=" + target.get("name")
                + "  mailbox=" + target.get("mailbox"));

        // Build A2A SendMessageRequest — message body is one or more Parts
        Message msg = new Message.Builder()
                .role(Message.Role.USER)
                .parts(new TextPart("你好，世界"))
                .build();
        SendMessageRequest request = new SendMessageRequest(
                null, new MessageSendParams(msg, null, null));

        // sendMessage returns msg_id; task_id is generated by Agent A (the executor)
        // and arrives with reply events, readable as context.taskId in onMessage
        System.out.println("[agent-b] sending task…");
        long msgId = agentB.sendMessage(target, request, bMailbox).join();
        System.out.println("[agent-b] sent  msg_id=" + msgId);

        System.out.println("[agent-b] waiting for reply…");
        boolean received = done.await(15, TimeUnit.SECONDS);
        if (!received) System.out.println("[agent-b] timed out waiting for reply");

        System.out.println("[agent-b] done.");
        agentB.close();
    }

    // ── Combined mode ──────────────────────────────────────────────────────────

    static void runBoth() throws Exception {
        Mq9A2AAgent agentA = Mq9A2AAgent.builder().server(SERVER).build();

        agentA.onMessage(
            (A2AContext ctx, EventQueue queue) ->
                queue.working(ctx)
                    .thenCompose(v -> {
                        String text = ctx.firstTextPart().orElse("");
                        System.out.println("[agent-a] translating: '" + text + "'");
                        return queue.artifact(ctx, "translation", "[translated] " + text);
                    })
                    .thenCompose(v -> queue.completed(ctx)),
            ConsumeOptions.builder()
                .groupName("demo.java.translator.workers")
                .deliver("earliest")
                .numMsgs(10)
                .maxWaitMs(500)
                .build()
        );

        agentA.connect().join();
        String aMailbox = agentA.createMailbox("demo.java.translator", 0).join();
        System.out.println("[agent-a] mailbox=" + aMailbox);

        Thread.sleep(500);
        runClient();

        agentA.close();
    }

    // ── Entry point ────────────────────────────────────────────────────────────

    public static void main(String[] args) throws Exception {
        String mode = args.length > 0 ? args[0] : "both";
        switch (mode) {
            case "agent"  -> runAgent();
            case "client" -> runClient();
            case "both"   -> runBoth();
            default -> {
                System.out.println("Usage: A2ADemo [agent|client|both]");
                System.exit(1);
            }
        }
    }
}

File: demo/python/a2a_demo.py
"""mq9 A2A Demo
=============
Demonstrates Mq9A2AAgent (server side) and Mq9A2AClient (client side)
communicating via the mq9 broker using the A2A protocol.

Run two terminals:

  Terminal 1 — start the agent:
    MQ9_SERVER=nats://demo.robustmq.com:4222 python a2a_demo.py agent

  Terminal 2 — send a task:
    MQ9_SERVER=nats://demo.robustmq.com:4222 python a2a_demo.py client

Or run both in the same process (agent runs in background):
    MQ9_SERVER=nats://demo.robustmq.com:4222 python a2a_demo.py both
"""

import asyncio
import os
import sys

from a2a.helpers import new_task_from_user_message, new_text_artifact, new_text_message
from a2a.server.agent_execution import RequestContext
from a2a.server.events import EventQueue
from a2a.types.a2a_pb2 import (
    AgentCard,
    AgentCapabilities,
    AgentSkill,
    SendMessageRequest,
    Role,
    TaskArtifactUpdateEvent,
    TaskState,
    TaskStatus,
    TaskStatusUpdateEvent,
)

from mq9.a2a import Mq9A2AAgent

SERVER = os.environ.get("MQ9_SERVER", "nats://demo.robustmq.com:4222")

AGENT_CARD = AgentCard(
    name="demo.a2a.translator",
    description="Multilingual translation agent. "
                "Send text + target_lang (en/zh/ja), get back translated text.",
    version="1.0.0",
    skills=[
        AgentSkill(
            id="translate",
            name="Translate text",
            description="Translates text between EN, ZH, JA, KO.",
        )
    ],
    capabilities=AgentCapabilities(streaming=True),
)


# ──────────────────────────────────────────────────────────────────────────────
# Agent side
# ──────────────────────────────────────────────────────────────────────────────

def make_agent() -> Mq9A2AAgent:
    agent = Mq9A2AAgent(server=SERVER)

    @agent.on_message
    async def handle(context: RequestContext, event_queue: EventQueue) -> None:
        task = context.current_task or new_task_from_user_message(context.message)
        await event_queue.enqueue_event(task)

        await event_queue.enqueue_event(
            TaskStatusUpdateEvent(
                task_id=context.task_id,
                context_id=context.context_id,
                status=TaskStatus(
                    state=TaskState.TASK_STATE_WORKING,
                    message=new_text_message("Translating…"),
                ),
            )
        )

        text = ""
        if context.message and context.message.parts:
            part = context.message.parts[0]
            if hasattr(part, "text"):
                text = part.text

        translated = f"[translated] {text}"

        await event_queue.enqueue_event(
            TaskArtifactUpdateEvent(
                task_id=context.task_id,
                context_id=context.context_id,
                artifact=new_text_artifact(name="translation", text=translated),
            )
        )
        await event_queue.enqueue_event(
            TaskStatusUpdateEvent(
                task_id=context.task_id,
                context_id=context.context_id,
                status=TaskStatus(state=TaskState.TASK_STATE_COMPLETED),
            )
        )
        print(f"[agent] processed: '{text}' → '{translated}'")

    return agent


async def run_agent() -> None:
    agent = make_agent()
    print(f"[agent] starting — name={AGENT_CARD.name}  server={SERVER}")
    await agent.connect()
    mailbox = await agent.register(AGENT_CARD)
    print(f"[agent] registered — mailbox={mailbox}")
    try:
        await asyncio.Event().wait()  # keep running until Ctrl+C
    except (KeyboardInterrupt, asyncio.CancelledError):
        pass
    finally:
        await agent.unregister()
        await agent.close()


# ──────────────────────────────────────────────────────────────────────────────
# Client side
# ──────────────────────────────────────────────────────────────────────────────

async def run_client() -> None:
    client = Mq9A2AAgent(server=SERVER)
    await client.connect()

    print("[client] discovering translation agents…")
    agents = await client.discover("translation agent")
    if not agents:
        print("[client] no agents found — is the agent running?")
        await client.close()
        return

    target = agents[0]
    print(f"[client] found agent: name={target.get('name')}  mailbox={target.get('mailbox')}")

    # ── 1. SendMessage ─────────────────────────────────────────────────────
    request = SendMessageRequest(
        message=new_text_message("你好，世界", role=Role.ROLE_USER)
    )

    print("\n[client] sending task…")
    msg_id = await client.send_message(target["mailbox"], request)
    print(f"[client] sent, msg_id={msg_id}")

    # ── 3. ListTasks ───────────────────────────────────────────────────────
    print("\n[client] ListTasks…")
    tasks = await client.list_tasks(target["mailbox"])
    print(f"[client] total tasks stored: {len(tasks)}")

    print("\n[client] done.")
    await client.close()


# ──────────────────────────────────────────────────────────────────────────────
# Combined mode: agent in background, client runs once then stops
# ──────────────────────────────────────────────────────────────────────────────

async def run_both() -> None:
    agent = make_agent()
    await agent.connect()
    mailbox = await agent.register(AGENT_CARD)
    print(f"[agent] registered — mailbox={mailbox}")

    await asyncio.sleep(1)

    await run_client()

    await agent.unregister()
    await agent.close()


# ──────────────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    mode = sys.argv[1] if len(sys.argv) > 1 else "both"
    match mode:
        case "agent":
            asyncio.run(run_agent())
        case "client":
            asyncio.run(run_client())
        case "both":
            asyncio.run(run_both())
        case _:
            print(__doc__)
            sys.exit(1)
